lxd/device/device_utils_unix_hotplug_events.go

```diff
@@ -0,0 +1,119 @@
+package device
+
+import (
+	"fmt"
+	"strconv"
+	"strings"
+	"sync"
+
+	deviceConfig "github.com/lxc/lxd/lxd/device/config"
+	"github.com/lxc/lxd/lxd/state"
+	log "github.com/lxc/lxd/shared/log15"
+	"github.com/lxc/lxd/shared/logger"
+)
+
+// UnixHotplugEvent represents the properties of a Unix hotplug device uevent.
+type UnixHotplugEvent struct {
+	Action string
+
+	Vendor  string
+	Product string
+
+	Path        string
+	Major       uint32
+	Minor       uint32
+	Subsystem   string
+	UeventParts []string
+	UeventLen   int
+}
+
+// unixHotplugHandlers stores the event handler callbacks for Unix hotplug events.
+var unixHotplugHandlers = map[string]func(UnixHotplugEvent) (*deviceConfig.RunConfig, error){}
+
+// unixHotplugMutex controls access to the unixHotplugHandlers map.
+var unixHotplugMutex sync.Mutex
+
+// unixHotplugRegisterHandler registers a handler function to be called whenever a Unix hotplug device event occurs.
+func unixHotplugRegisterHandler(instance Instance, deviceName string, handler func(UnixHotplugEvent) (*deviceConfig.RunConfig, error)) {
+	unixHotplugMutex.Lock()
+	defer unixHotplugMutex.Unlock()
+
+	// Null delimited string of project name, instance name and device name.
+	key := fmt.Sprintf("%s\000%s\000%s", instance.Project(), instance.Name(), deviceName)
```

Comment:
@brauner I can answer that one, its been copied from `unixRegisterHandler` which I added as part of the device framework for existing hotplugging of unix devices using inotify. https://github.com/lxc/lxd/blob/master/lxd/device/device_utils_unix_events.go#L44

I chose null termination because underscores are not currently disallowed in all of those 3 variables AFAIK.